trunk/src/app/srs_app_config.cpp

```diff
@@ -69,14 +69,17 @@ const char* _srs_version = "XCORE-" RTMP_SIG_SRS_SERVER;
 #define SRS_OVERWRITE_BY_ENV_FLOAT_SECONDS(key) if (!srs_getenv(key).empty()) return srs_utime_t(::atof(srs_getenv(key).c_str()) * SRS_UTIME_SECONDS)
 #define SRS_OVERWRITE_BY_ENV_FLOAT_MILLISECONDS(key) if (!srs_getenv(key).empty()) return srs_utime_t(::atof(srs_getenv(key).c_str()) * SRS_UTIME_MILLISECONDS)
 #define SRS_OVERWRITE_BY_ENV_DIRECTIVE(key) {                                 \
-        static SrsConfDirective* dir = NULL;                                  \
+        SrsConfDirective* dir = env_dirs->get(key);                           \
```

Comment:
Why change local variable to member field?

Reply:
the `static` var is to workaround the `memory leak`.
the problem of `static` is that the env cannot set twice in UT. once the env is parsed, the `static` dir set value, this macro will not parse any new values again, else a new process started, that's a problem for the UT.

If a better UT is our target, the solution is to add another root `directive`, env_dirs, to store the env parsed directives. Then this macro can parse new env value for each `SrsConfig` instance once.

Reply:
There is a logical error here: using static creates a variable in each function, while you are sharing a variable in the config. This will cause incompatibility with the previous logic.

For example, if you call get_vhost_on_connect first and then call get_vhost_on_close, the latter will retrieve the data from the former.

Reply:
No, I don't think so. 

`SRS_OVERWRITE_BY_ENV_DIRECTIVE` is a macro, which will expend before compiling, while the `env_dirs` is the `dir`'s container, which is managed by `SrsConfig`.

The `dir` variable is shared among function, that's true, because its container `env_dirs` is the member variable of `SrsConfig`, whose member function can access it. But `get_vhost_on_connect` and `get_vhost_on_close` only create or retrieve the `dir` by its key, in this case, they are `SRS_VHOST_HTTP_HOOKS_ON_CONNECT` and `SRS_VHOST_HTTP_HOOKS_ON_CLOSE`.

Almost same logic compared with the `static` solution, but the `get_vhost_on_connect` will retrieve the dir from env for each `SrsConfig` instance.

Reply:
Yep, you are correct, no problem.